src/main/java/org/opensearch/knn/plugin/script/KNNPainlessScriptUtils.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.plugin.script;
+
+import lombok.NonNull;
+import org.opensearch.knn.index.SpaceType;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utility class for k-NN scoring functions used in Painless scripts.
+ * Provides late interaction scoring functionality for ColBERT-style token-level matching.
+ */
+public class KNNPainlessScriptUtils {
+
+    /**
+     * Calculates the late interaction score between query vectors and document vectors using inner product.
+     * For each query vector, finds the maximum similarity with any document vector and sums these maxima.
+     * This implements a ColBERT-style late interaction pattern for token-level matching.
+     *
+     * @param queryVectors List of query vectors
+     * @param docFieldName Name of the field in the document containing vectors
+     * @param doc Document source as a map
+     * @return Sum of maximum similarity scores
+     */
+    public static float lateInteractionScore(
+        @NonNull final List<List<Number>> queryVectors,
+        @NonNull final String docFieldName,
+        @NonNull final Map<String, Object> doc
+    ) {
+        return lateInteractionScore(queryVectors, docFieldName, doc, SpaceType.INNER_PRODUCT.getValue());
+    }
+
+    /**
+     * Calculates the late interaction score between query vectors and document vectors with specified similarity metric.
+     * For each query vector, finds the maximum similarity with any document vector and sums these maxima.
+     * This implements a ColBERT-style late interaction pattern for token-level matching.
+     *
+     * @param queryVectors List of query vectors
+     * @param docFieldName Name of the field in the document containing vectors
+     * @param doc Document source as a map
+     * @param spaceType Space type for similarity calculation: "innerproduct", "cosinesimil", "l2", "l1", "linf"
+     * @return Sum of maximum similarity scores
+     */
+    @SuppressWarnings("unchecked")
+    public static float lateInteractionScore(
+        @NonNull final List<List<Number>> queryVectors,
+        @NonNull final String docFieldName,
+        @NonNull final Map<String, Object> doc,
+        @NonNull final String spaceType
```

Comment:
does these NonNull checks valid here? In my understanding the painless script will directly call these functions, and it can happen that docFieldName is null and can lead to NPE due to @NonNull annotation. It will be better if you can add a validation utility here which can validate all the fields and then throw a user friendly exception and message to user

Reply:
removed NonNull and also add validation method